File: src/colibri/tools/shell_policy.py
from __future__ import annotations

import os
import shlex


def denied_shell_executable(command: str, deny: set[str]) -> str | None:
    for executable in shell_executables(command):
        name = os.path.basename(executable)
        if executable in deny or name in deny:
            return executable
    return None


def first_shell_executable(command: str) -> str | None:
    return next(iter(shell_executables(command)), None)


def shell_executables(command: str) -> list[str]:
    executables: list[str] = []
    for segment in shell_command_segments(command):
        try:
            argv = shlex.split(segment)
        except ValueError:
            continue
        executable = _first_executable_token(argv)
        if executable is not None:
            executables.append(executable)
    return executables


def has_dangerous_shell_features(command: str) -> bool:
    if _has_unquoted_nested_shell_syntax(command):
        return True
    for segment in shell_command_segments(command):
        try:
            argv = shlex.split(segment)
        except ValueError:
            continue
        executable = _first_executable_token(argv)
        if executable in {"eval", "source", "."}:
            return True
    return False


def shell_command_segments(command: str) -> list[str]:
    segments: list[str] = []
    buffer: list[str] = []
    quote: str | None = None
    escaped = False
    index = 0
    while index < len(command):
        char = command[index]
        if escaped:
            buffer.append(char)
            escaped = False
            index += 1
            continue
        if char == "\\" and quote != "'":
            buffer.append(char)
            escaped = True
            index += 1
            continue
        if quote is not None:
            buffer.append(char)
            if char == quote:
                quote = None
            index += 1
            continue
        if char in {"'", '"'}:
            quote = char
            buffer.append(char)
            index += 1
            continue
        if char == "\n" or char == ";":
            _push_segment(segments, buffer)
            index += 1
            continue
        if char == "|" and (index + 1 >= len(command) or command[index + 1] != "|"):
            _push_segment(segments, buffer)
            index += 1
            continue
        if char in {"&", "|"} and index + 1 < len(command) and command[index + 1] == char:
            _push_segment(segments, buffer)
            index += 2
            continue
        if char == "&" and not _is_ampersand_redirection(command, index):
            _push_segment(segments, buffer)
            index += 1
            continue
        buffer.append(char)
        index += 1
    _push_segment(segments, buffer)
    return segments


def _has_unquoted_nested_shell_syntax(command: str) -> bool:
    quote: str | None = None
    escaped = False
    index = 0
    while index < len(command):
        char = command[index]
        if escaped:
            escaped = False
            index += 1
            continue
        if char == "\\" and quote != "'":
            escaped = True
            index += 1
            continue
        if quote is not None:
            if char == quote:
                quote = None
            index += 1
            continue
        if char in {"'", '"'}:
            quote = char
            index += 1
            continue
        if command.startswith("$(", index) or char in {"`", "("}:
            return True
        if command.startswith("<<", index):
            return True
        index += 1
    return False


def _is_ampersand_redirection(command: str, index: int) -> bool:
    previous_char = command[index - 1] if index > 0 else ""
    next_char = command[index + 1] if index + 1 < len(command) else ""
    return previous_char == ">" or next_char == ">"


def _push_segment(segments: list[str], buffer: list[str]) -> None:
    segment = "".join(buffer).strip()
    if segment:
        segments.append(segment)
    buffer.clear()


def _first_executable_token(argv: list[str]) -> str | None:
    for token in argv:
        if _is_env_assignment(token):
            continue
        executable = _strip_inline_redirection(token)
        if executable:
            return executable
    return None


def _is_env_assignment(token: str) -> bool:
    if "=" not in token:
        return False
    name, _, _value = token.partition("=")
    return bool(name) and (name[0].isalpha() or name[0] == "_") and all(ch.isalnum() or ch == "_" for ch in name)


def _strip_inline_redirection(token: str) -> str:
    for marker in ("<", ">"):
        if marker in token:
            token = token.split(marker, 1)[0]
    return token

File: colibri-rust/src/shell_policy.rs
use std::path::Path;

pub(crate) fn denied_shell_executable(command: &str, deny: &[String]) -> Option<String> {
    for executable in shell_executables(command) {
        let executable_name = Path::new(&executable)
            .file_name()
            .map(|name| name.to_string_lossy().to_string())
            .unwrap_or_else(|| executable.clone());
        if deny
            .iter()
            .any(|denied| denied == &executable || denied == &executable_name)
        {
            return Some(executable);
        }
    }
    None
}

pub(crate) fn first_shell_executable(command: &str) -> Option<String> {
    shell_executables(command).into_iter().next()
}

pub(crate) fn shell_executables(command: &str) -> Vec<String> {
    shell_command_segments(command)
        .into_iter()
        .filter_map(|segment| {
            shell_words::split(&segment)
                .ok()
                .and_then(|argv| first_executable_token(&argv))
        })
        .collect()
}

pub(crate) fn has_dangerous_shell_features(command: &str) -> bool {
    if has_unquoted_nested_shell_syntax(command) {
        return true;
    }
    shell_command_segments(command).into_iter().any(|segment| {
        shell_words::split(&segment)
            .ok()
            .and_then(|argv| first_executable_token(&argv))
            .is_some_and(|executable| matches!(executable.as_str(), "eval" | "source" | "."))
    })
}

pub(crate) fn shell_command_segments(command: &str) -> Vec<String> {
    let chars: Vec<char> = command.chars().collect();
    let mut segments = Vec::new();
    let mut buffer = String::new();
    let mut quote: Option<char> = None;
    let mut escaped = false;
    let mut index = 0;
    while index < chars.len() {
        let ch = chars[index];
        if escaped {
            buffer.push(ch);
            escaped = false;
            index += 1;
            continue;
        }
        if ch == '\\' && quote != Some('\'') {
            buffer.push(ch);
            escaped = true;
            index += 1;
            continue;
        }
        if let Some(quote_char) = quote {
            buffer.push(ch);
            if ch == quote_char {
                quote = None;
            }
            index += 1;
            continue;
        }
        if ch == '\'' || ch == '"' {
            quote = Some(ch);
            buffer.push(ch);
            index += 1;
            continue;
        }
        if ch == '\n' || ch == ';' {
            push_shell_segment(&mut segments, &mut buffer);
            index += 1;
            continue;
        }
        if ch == '|' && (index + 1 >= chars.len() || chars[index + 1] != '|') {
            push_shell_segment(&mut segments, &mut buffer);
            index += 1;
            continue;
        }
        if (ch == '&' || ch == '|') && index + 1 < chars.len() && chars[index + 1] == ch {
            push_shell_segment(&mut segments, &mut buffer);
            index += 2;
            continue;
        }
        if ch == '&' && !is_ampersand_redirection(&chars, index) {
            push_shell_segment(&mut segments, &mut buffer);
            index += 1;
            continue;
        }
        buffer.push(ch);
        index += 1;
    }
    push_shell_segment(&mut segments, &mut buffer);
    segments
}

fn has_unquoted_nested_shell_syntax(command: &str) -> bool {
    let chars: Vec<char> = command.chars().collect();
    let mut quote: Option<char> = None;
    let mut escaped = false;
    let mut index = 0;
    while index < chars.len() {
        let ch = chars[index];
        if escaped {
            escaped = false;
            index += 1;
            continue;
        }
        if ch == '\\' && quote != Some('\'') {
            escaped = true;
            index += 1;
            continue;
        }
        if let Some(quote_char) = quote {
            if ch == quote_char {
                quote = None;
            }
            index += 1;
            continue;
        }
        if ch == '\'' || ch == '"' {
            quote = Some(ch);
            index += 1;
            continue;
        }
        if ch == '`'
            || ch == '('
            || (ch == '$' && index + 1 < chars.len() && chars[index + 1] == '(')
        {
            return true;
        }
        if ch == '<' && index + 1 < chars.len() && chars[index + 1] == '<' {
            return true;
        }
        index += 1;
    }
    false
}

fn is_ampersand_redirection(chars: &[char], index: usize) -> bool {
    let previous = index
        .checked_sub(1)
        .and_then(|previous| chars.get(previous));
    let next = chars.get(index + 1);
    previous == Some(&'>') || next == Some(&'>')
}

fn push_shell_segment(segments: &mut Vec<String>, buffer: &mut String) {
    let segment = buffer.trim();
    if !segment.is_empty() {
        segments.push(segment.to_string());
    }
    buffer.clear();
}

fn first_executable_token(argv: &[String]) -> Option<String> {
    for token in argv {
        if is_env_assignment(token) {
            continue;
        }
        let executable = strip_inline_redirection(token);
        if !executable.is_empty() {
            return Some(executable);
        }
    }
    None
}

fn is_env_assignment(token: &str) -> bool {
    let Some((name, _value)) = token.split_once('=') else {
        return false;
    };
    let mut chars = name.chars();
    let Some(first) = chars.next() else {
        return false;
    };
    (first.is_ascii_alphabetic() || first == '_')
        && chars.all(|ch| ch.is_ascii_alphanumeric() || ch == '_')
}

fn strip_inline_redirection(token: &str) -> String {
    token
        .split(['<', '>'])
        .next()
        .unwrap_or_default()
        .to_string()
}
